File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/AuditoriumSeatingArrangement.java
package org.weaveit.seatingplacesuggestions;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public record AuditoriumSeatingArrangement(Map<String, Row> rows) {

    public AuditoriumSeatingArrangement {
        rows = Collections.unmodifiableMap(new LinkedHashMap<>(rows));
    }

    public SeatingOption suggestSeatingOptionFor(int partyRequested, PricingCategory pricingCategory) {
        for (Row row : rows.values()) {
            var seatingOption = row.suggestSeatingOption(partyRequested, pricingCategory);

            if (seatingOption.matchExpectation()) {
                return seatingOption;
            }
        }

        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public AuditoriumSeatingArrangement allocate(List<SeatingPlace> seatsToAllocate) {
        Map<String, Row> updatedRows = new LinkedHashMap<>();
        for (var entry : rows.entrySet()) {
            Row updatedRow = entry.getValue().allocate(seatsToAllocate);
            updatedRows.put(entry.getKey(), updatedRow);
        }
        return new AuditoriumSeatingArrangement(updatedRows);
    }
}

File: TheaterSuggestions/Kotlin/ScreeningSeatingArrangementSuggestions/src/main/kotlin/org/weaveit/seatingplacesuggestions/AuditoriumSeatingArrangement.kt
package org.weaveit.seatingplacesuggestions

data class AuditoriumSeatingArrangement private constructor(
    private val _rows: Map<String, Row>
) {
    val rows: Map<String, Row> get() = _rows

    companion object {
        operator fun invoke(rows: Map<String, Row>): AuditoriumSeatingArrangement =
            AuditoriumSeatingArrangement(LinkedHashMap(rows))
    }

    fun suggestSeatingOptionFor(partyRequested: Int, pricingCategory: PricingCategory): SeatingOptionIsSuggested {
        for (row in rows.values) {
            val seatingOptionSuggested = row.suggestSeatingOption(partyRequested, pricingCategory)

            if (seatingOptionSuggested.matchExpectation()) {
                return seatingOptionSuggested
            }
        }

        return SeatingOptionIsNotAvailable(partyRequested, pricingCategory)
    }

    fun allocate(seatsToAllocate: List<SeatingPlace>): AuditoriumSeatingArrangement {
        val updatedRows = rows.mapValues { (_, row) ->
            row.allocate(seatsToAllocate)
        }
        return AuditoriumSeatingArrangement(LinkedHashMap(updatedRows))
    }
}
